librad/src/net/protocol.rs

```diff
@@ -194,90 +206,126 @@ where
     }
 
     /// Open a QUIC stream which is upgraded to expect the git protocol
-    pub async fn open_git(
+    ///
+    /// If no connection to the given peer is currently active, `addr_hints`
+    /// will be used to attempt to establish a new connection. `addr_hints` must
+    /// be finite -- we use the [`IntoIterator`] trait bound for notational
+    /// convenience (one can pass [`None`], for example).
+    pub async fn open_git<Addrs>(
         &self,
         to: &PeerId,
-    ) -> Result<Upgraded<quic::Stream, upgrade::Git>, Error> {
-        self.open_stream(to, upgrade::Git).await
+        addr_hints: Addrs,
+    ) -> Result<Upgraded<quic::Stream, upgrade::Git>, Error>
+    where
+        Addrs: IntoIterator<Item = SocketAddr>,
```

Comment:
FYI there's also https://doc.rust-lang.org/std/net/trait.ToSocketAddrs.html#tymethod.to_socket_addrs which is handy.

Reply:
This is fallible, though, I think it calls `gethostbyname` if it’s not an IP address. We can keep it pure, because we know we have `SocketAddr`s (or the caller can resolve).

Reply:
True that, it can convert from all sorts of inputs.